Add 'getOrPut' operation for maps to the new base language extensions

dalbrecht1 opened the pull request:
## Name
getOrPut for maps (`map[key, (k) => fallback]`)

## Operation Type
Expression

## Use Case
When you use maps in MPS to collect values associated with keys, and if you aren't sure if a key you would use to access the map is already associated with anything, then it can be a bit verbose to either check for the presence of a value first, or do a cast to the underlying Java-types in order to use convenience methods that already exist on that level.

To assist in these situations the mentioned Java-Map-method - to query a key and if absent initialize the mapping with a provided value and return either the already existing value or the newly added one - was added to the MPS-collection abstraction level in form of an alternative MapElement (the `map[key]` syntax), that makes this Java-method accessible from within pure MPS base Language code.

## Comments
This PR also supplies two intentions that help migrating from one style of map-access to the other. But in the case that you want to convert a classical 'get' to the 'getOrPut' syntax, then the intention will not be able to pre-set the type of the closure for MPS-internal reasons I wasn't able to figure out and solve.

dalbrecht1 committed:
added 'getOrPut' operation for maps

alamqadem committed:
update build script dependencies

alamqadem committed:
add example with lists to the sandbox

alamqadem reviewed (changes requested):
- [x] I've tested it and it works as expected, I've pushed some commits to update the build script and add another small example to the sandbox;
- [ ] for the generator why not using the approach of calling a static method in the runtime? What the `generic lvalue` expression does is not really clear to me.

dalbrecht1 commented:
> for the generator why not using the approach of calling a static method in the runtime? What the `generic lvalue` expression does is not really clear to me.

I think I would still figuratively die on that hill that this was the better design choice, but I agree and admit, that I did a poor job of documenting the design decisions here and explain the way I understand the generator code I "got inspired from".

I could have designed this operation like any other extension method and call it ".getOrPut()" or something. But that didn't align really with what was there already as the "get" operation isn't an extension either but "special syntax" (the `map[key]` syntax). So I thought why not just extend the given special syntax get-operation and also align in its editor representation. And that's what I did.

Now, this operation not only has special editor syntax, but also is different in its generator choice. It generates code using the lvalue intermediate step. This structure is used (as far as my understanding goes) for any values (used in assignments or "calculation"), that can be used on the right hand side (as any other expression), but on top ALSO supports being the target of an assignment, hence, going on the LEFT hand side (therefore, lvalue).

To accomplish that, the lvalue construct needs to know, how to further handle this particular expression element, specifically to the use case (or in MPS terms "link") it is used in. If it appears as a value expression on the right hand side, it needs to know what value it represents and if it is used on the left hand side, then it has to know how to consume values for storage/update.

That also means, that my new "operator" is usable as a left hand side expression, where the "put" aspect of my operation WILL NEVER BE TRIGGERED. So, using this for a left hand side practically DOES NOT MAKE SENSE. But that was the consequence of extending this existing concept. But given that the put-closure also isn't ALWAYS invoked if used as a right hand side expression, I'm fine with someone using that as a left hand side, where there then is a guarantee that the closure isn't used. One could however add a warning in case this new concept is used as left hand side, to nudge the user towards using the shorter default get notation. (Tell me what you think and I will add it if you agree with my reasoning here of having a checking rule.)

But I'd like to keep the lvalue here in the generator, as that aligns with how the "parent concept" is generated, so for the unlikely case that the generator might make assumptions about "MapGet" turns into "lvalue", that then my construct doesn't break that assumption. And I want to keep the "MapGet" as my parent concept, as I otherwise would have to re-implement a bunch if its aspects again. And reuse is king here.

Hope, that helped understanding the reasoning and the implementation.

dalbrecht1 committed:
moved type-inference out of execute, as typesystem is only available in preparation steps

dalbrecht1 committed:
further improvements to the getOrPut operation

- based on Sergej's feedback there is now an alternative syntax (a method) to trigger that functionality
- still, based on the fact that the Java method-name in my eyes is imprecise I stuck with my name for now
- the data flow aspect get improved, previously I had sort of guessed, now I actually looked up, how that works
